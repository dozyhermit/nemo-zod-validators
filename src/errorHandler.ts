import { z } from 'zod';
import type { ZodError } from './types';

export const DEFAULT_ERROR_MESSAGE = 'Validation has failed';

// TECHNICAL DEBT: remove .flatten(), use z.flattenError instead
const getErrors = (
  error: ZodError['error'],
  key: keyof z.core.$ZodFlattenedError<string, unknown>
) => {
  if (typeof error?.flatten !== 'function') {
    return undefined;
  }

  if (Object.values(error?.flatten()?.[key] || {}).length <= 0) {
    return undefined;
  }

  return error.flatten()[key];
};

export const errorHandlerWithSchema = ({ error }: ZodError) =>
  Response.json(
    {
      message: getErrors(error, 'fieldErrors') || DEFAULT_ERROR_MESSAGE,
    },
    { status: 422 }
  );

export const errorHandlerWithSchemaFormErrors = ({ error }: ZodError) =>
  Response.json(
    {
      message: getErrors(error, 'formErrors') || DEFAULT_ERROR_MESSAGE,
    },
    { status: 422 }
  );

export const errorHandlerWithTransform = () =>
  Response.json(
    { message: 'Invalid request', success: false },
    {
      status: 400,
    }
  );
