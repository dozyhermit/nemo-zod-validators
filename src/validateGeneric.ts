import { NextResponse } from 'next/server';
import { errorHandlerWithSchema } from './errorHandler';
import { ValidateWithSchema, type ValidateBaseReturnType } from './types';

type ValidateGeneric<T> = { data: T } & ValidateWithSchema<T>;

/**
 * Validates data with a zod schema using safeParse.
 *
 * @param {T} data The data to be validated
 * @param {z.ZodObject} schema The schema doing the validating
 * @param {Function} errorHandler A custom error function for custom error return results
 */
export const validateGeneric = <T>({
  data,
  schema,
  errorHandler,
}: ValidateGeneric<T>): ValidateBaseReturnType => {
  const validationResult = schema.safeParse(data as T);

  if (validationResult.error) {
    if (errorHandler) {
      return errorHandler(validationResult);
    }

    return errorHandlerWithSchema(validationResult);
  }

  return NextResponse.next();
};
