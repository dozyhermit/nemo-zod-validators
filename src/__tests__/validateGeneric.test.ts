import { validateGeneric } from '../validateGeneric';
import { errorHandler } from './mocks/errorHandler';
import {
  type SchemaType,
  schema,
  schemaWithIntentionallyBrokenSafeParse,
} from './mocks/schema';

describe('validateGeneric', () => {
  test('passes validation with correct data', () => {
    const result = validateGeneric<SchemaType>({
      data: { Number: 123, String: 'Hello', AdvancedString: '123456' },
      schema,
    });

    expect(result.status).toBe(200);
  });

  test('triggers validation error with incorrect data', async () => {
    const result = validateGeneric<SchemaType>({
      data: { Number: 123, String: 'Hello', AdvancedString: 'World' },
      schema,
    });

    expect(await result.json()).toMatchSnapshot();
  });

  test('triggers validation error with incorrect data and uses DEFAULT_ERROR_MESSAGE', async () => {
    const result = validateGeneric<SchemaType>({
      data: { Number: 123, String: 'Hello', AdvancedString: 'World' },
      // @ts-expect-error intentionally broken schema
      schema: schemaWithIntentionallyBrokenSafeParse,
    });

    expect(await result.json()).toMatchSnapshot();
  });

  describe('with custom error handler', () => {
    test('passes validation with correct data', () => {
      const result = validateGeneric<SchemaType>({
        data: { Number: 123, String: 'Hello', AdvancedString: '123456' },
        schema,
        errorHandler,
      });

      expect(result.status).toBe(200);
    });

    test('triggers validation error with incorrect data', async () => {
      const result = validateGeneric<SchemaType>({
        data: { Number: 123, String: 'Hello', AdvancedString: 'World' },
        schema,
        errorHandler,
      });

      expect(await result.json()).toMatchSnapshot();
    });
  });
});
