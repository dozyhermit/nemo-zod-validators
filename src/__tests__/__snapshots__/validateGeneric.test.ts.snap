// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`validateGeneric triggers validation error with incorrect data 1`] = `
{
  "message": {
    "AdvancedString": [
      "Invalid AdvancedString regex",
    ],
  },
}
`;

exports[`validateGeneric triggers validation error with incorrect data and uses DEFAULT_ERROR_MESSAGE 1`] = `
{
  "message": "Validation has failed",
}
`;

exports[`validateGeneric with custom error handler triggers validation error with incorrect data 1`] = `
{
  "message": "Invalid validation!",
}
`;
